### PR TITLE
Add end_training/destroy_pg to everything and unpin numpy

### LINKED ISSUE #3027: Support NumPy 2.0 and Greater.
Please excuse the blank template, but the only template publicly available for accelerate is a bug template which is not applicable here.

Currently accelerate requires numpy less than version 2.0. This has resulted in numpy being forcibly downgraded in some environments such as what is happening in https://github.com/huggingface/transformers/issues/31802. This issue was closed with Hugging Face employees recommending that a ticket for NumPy 2.0 support be opened in accelerate.

If you could support NumPy version 2.0 (or whatever the latest version of NumPy is) I and the rest of the accelerate and Hugging Face user base would be grateful.

### PR DESCRIPTION
**Pull request by muellerzr**
# What does this PR do?

Some more spring cleaning. Adds `end_training`/`state.destroy_process_group()` everywhere we need to in our examples, and unpins numpy as things look to be fine from what I can tell

Fixes #3027 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/accelerate/blob/main/CONTRIBUTING.md#submitting-a-pull-request-pr),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/accelerate/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/accelerate/tree/main/docs#writing-documentation---specification).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@SunMarc

### CODE REVIEW

**Review by SunMarc — APPROVED** [review]
Nice !

File: setup.py [hunk 1]
```diff
@@ -70,7 +70,7 @@
     },
     python_requires=">=3.8.0",
     install_requires=[
-        "numpy>=1.17,<2.0.0",
+        "numpy>=1.17,<3.0.0",
```
[the remark this answers is not in the document]
Comment (muellerzr) [reply setup.py:73]: We want to mark a top, similar to here, so we know what issues we have :)

**Comment by BenjaminBossan** [comment]
@muellerzr Do you know if numpy 2.0 now also works with PyTorch < v2.4?

**Comment by muellerzr** [comment]
In my very quick test, it looks like yes it's fine (ran the accelerate tests from 2.3 and 2.2

**Comment by BenjaminBossan** [comment]
Nice, good to know. It sounded like older releases would not work

> Older releases not supporting NumPy 2.0 is expected and will not be fixed as per the usual policy we have on old releases patching.

`https://github.com/pytorch/pytorch/issues/ 128860`

**Comment by muellerzr** [comment]
Since we don't have a "base" requirement for 2.0, we don't force people to upgrade.

**Pull request merged**